[Streamed Orderbook] Read block timestamps in batches

fleupold opened the pull request:
This PR improves the long cold start time when using the event based orderbook, by batching the requests for block timestamps. 
With the current business logic, we need to get the timestamp for each block that emitted a relevant event. On rinkeby alone this results in a few thousand requests that we send to the FullNode. By batching these calls into a few single requests we can significantly decrease roundtrip time and network overhead.

On my local machine (4G speed) I can now sync the event based orderbook on rinkeby in less than a minute (before it would take more than 20 miuntes) - **a speedup of 20x.**

Future changes to the business logic might make it so we don't have to fetch historic timestamps anymore (cf. discussion in #760), however even for our incremental fetching it can be beneficial to batch the timestamp requests.

My approach is to change the `MemoizingBlockTimestampReader` to contain a full map instead of just the last block's timestamp and furthermore make it "pre-heatable" by querying a bunch of block timestamps in advance. This way, it will instantly return the queried block hashes from memory (at least for now for the past events).

### Test Plan

run it using:

`cargo run -- --network-id=4 --node-url=https://blue-parity.rinkeby.gnosisdev.com --use-shadowed-orderbook true --primary_orderbook EventBased`

and see that we have "Finished applying past events" much earlier than before.

fleupold committed:
[Streamed Orderbook] `Read block timestamps in batches

josojo committed:
solver bump (#790)

dependabot-preview[bot] committed:
Bump criterion from 0.3.1 to 0.3.2

Bumps [criterion](https://github.com/bheisler/criterion.rs) from 0.3.1 to 0.3.2.
- [Release notes](https://github.com/bheisler/criterion.rs/releases)
- [Changelog](https://github.com/bheisler/criterion.rs/blob/master/CHANGELOG.md)
- [Commits](https://github.com/bheisler/criterion.rs/compare/0.3.1...0.3.2)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

e00E commented on driver/src/orderbook/streamed/block_timestamp_reading.rs:
```diff
@@ -22,34 +36,83 @@ impl BlockTimestampReading for Web3 {
+        }
+    }
+
+    pub fn prepare_cache(&mut self, block_hashes: HashSet<H256>) -> BoxFuture<Result<()>> {
+        async move {
+            self.cache
+                .extend(self.inner.block_timestamps(block_hashes).await?);
+            Ok(())
         }
```
Here we are querying all timestamps from the node even those that are already in the cache. Doesn't matter in practice because it's only called for past_events where we don't have any timestamps.

fleupold replied:
Very good point. Will make it so that it filters out the ones it already has.

nlordell reviewed (approved):
:guitar: :metal:

nlordell commented on driver/src/orderbook/streamed/block_timestamp_reading.rs:
```diff
@@ -22,34 +36,83 @@ impl BlockTimestampReading for Web3 {
+
+/// A cache for the block timestamp.
 #[derive(Debug)]
 pub struct MemoizingBlockTimestampReader<T> {
```
naming nit: This should probably be called `CachedBlockTimestampReader` (I reviewed the original PR before the name change) as I would describe what this is doing as caching more than memoizing.

fleupold committed:
Don't refetch previously fetched hashes

fleupold committed:
rename

fleupold committed:
[Easy] Log error chain or streamed orderbook exit (#789)

Three small changes to the way we exit the updating orderbook in case of an error:

1. Print the entire error stack trace not just the top level (by using the Debug print of anyhow)
2. Sleep for 1s before exiting to allow the logs to be flushed (I tried synchronously flushing them but that did not work)
3. Use process::exit instead of abort, as suggested by @nlordell (not feeling strongly about this)

### Test Plan

When working with the streamed orderbook and runnig into an error, make sure we can read the error reason and error chain.

fleupold committed:
[Streamed Orderbook] Fix batchId off-by-one error (#786)

This is a ~pretty~ extremely hacky short-term fix for #785 with the goal to keep on getting more consistency results and elminate known issues.

To summarize, `orderbook_for_batch` currently returns the balances that are valid in `batch - 1` and the orders that are valid in `batch`. This PR changes it to also return the orders that are valid in `batch-1`. Then, we need to increment the batchID we get passed in from `OrderbookReading` as this is already the index of the batch we are trying to solve (and for which we want the orderbook).

I also renamed the variable in the OrderbookReading trait to make it more explicit which batch this ID is for.

I'm happy to make/accept a cleaner fix if you guys have an idea.

### Test Plan

1. The first bug was that we would get errors if an event is received in the first 30s of a batch (since this would advance last_batch_id and then result assertion errors that we are querying on a past batch id). I tested locally that this is no longer the case (https://rinkeby.etherscan.io/tx/0x7556fe7b1ddfcb51d1b72cbc67c3060bad9dc5cae3639cd540f74c363e8c4942)
2. When creating an order and a deposit in the same batch the order would get correctly included by the streamed orderbook, but the deposit would be ignored. I tested locally that this is no longer the case (tx from above + deposit in the same batch https://rinkeby.etherscan.io/tx/0x63f94ce7cd43aaf2295da66315fffc2dffd43612f1308ec5dd3725717bf6707c)

fleupold committed:
Implement BatchTransport via Http (#787)

This PR implements the BatchTransport trait for our own HttpTransport implementation. Instead of sending a single JSON object containing the method, parms, id, etc, we send a list of multiple objects (potentially for different methods).

In the success case the result is a list of response objects one for each request.

### Test Plan

Using the new transport for fetching multiple block timestamps in a single request in the next PR.
